core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -263,7 +272,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         val workOffers = activeExecutors.map {
           case (id, executorData) =>
             new WorkerOffer(id, executorData.executorHost, executorData.freeCores,
-              Some(executorData.executorAddress.hostPort))
+              Some(executorData.executorAddress.hostPort), executorData.availableResources.toMap)
```

Comment:
we are passing in an immutable map here, which we should be, but then I think you are relying on that tasksetmanager to acquire and set the reserved, which then below the assignAddresses is supposed to account for.  That isn't going to work since its a new map we pass in.  Or at least there will never be reserved ones here in  executorData.availableResources. I guess other then that it will work as far as idle and assigned.
In the current usage I don't think we need reservedAddresses. It didn't really change our original comments about the place its doing the accounting in TaskSchedulerImpl vs TaskSetManager.  So lets just add a comment there

Reply:
Actually this works because `executorData.availableResources.toMap` just turns `Map[String, ExecutorResourceInfo]` into an immutable map, so we are still manifesting on the original ExecutorResourceInfo object. I updated a test case in `CoarseGrainedSchedulerBackendSuite` to show the case.

Another thing is, inside `CoarseGrainedSchedulerBackend.launchTasks()`, we may skip the task launch (when the size of serializedTask exceeds the max size allowed), under this case we shall release the resource addresses linked to the task directly (otherwise these resource addresses won't get returned). Under this case I prefer to release the `reservedAddresses` instead of `allocatedAddresses` because the addresses are not actually allocated.

Reply:
Oh sorry you are right, I misread, and that is actually worse I think.  I don't think we want to allow the taskScheduler to be modifying a data structure owned by the CoarseGrainedSchedulerBackend.  At this point free cores is handled separately. We create the Worker offers that has its own book keeping separate from the executorDataMap in CoarseGrainedSchedulerBacked. The task scheduler can do whatever it wants with the WorkerOffers and then the CoarseGrainedSchedulerBackend is responsible for updating the  executorDataMap. 

 If they are kept totally separate you don't have to worry about the executorDataMap being wrong if serializedTask is to large or if some other thing happens before the launchTasks.  If you keep them together you have to be very sure any thing that fails after reservation unreserves them, which I think is more of a maintenance issue.   If you miscount in one iteration of the task scheduler loop dealing with WorkerOffers it is kind of ok in the sense the next makeOffers call will fix it up and have the right data since it will be pulled from the separate executorDataMap.

So I think we should make a deep copy of that to send into the WorkerOffer.  normally I wouldn't expect much to be in there so it shouldn't be to expensive to do that.  But it does get called a lot if you have a lot of executors so it might be interesting to see what the overhead of that is when you have thousands of executors.  I'll try to look at it some more tomorrow.

Reply:
Discuss with @jiangxb1987 offline. Instead of a deep copy that might be costly, he will provide a read-only view in work offers.